File: go/src/617.合并二叉树.go
/*
 * @lc app=leetcode.cn id=617 lang=golang
 *
 * [617] 合并二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func mergeTrees(t1 *TreeNode, t2 *TreeNode) *TreeNode {
	if t1 == nil {
		return t2
	}
	if t2 == nil {
		return t1
	}
	dfs(t1, t2)
	return t1
}

func dfs(t1 *TreeNode, t2 *TreeNode) {
	t1.Val += t2.Val
	if t1.Left != nil && t2.Left != nil {
		dfs(t1.Left, t2.Left)
	} else if t1.Left == nil {
		t1.Left = t2.Left
	}

	if t1.Right != nil && t2.Right != nil {
		dfs(t1.Right, t2.Right)
	} else if t1.Right == nil {
		t1.Right = t2.Right
	}
}
// @lc code=end

File: rust/src/617.合并二叉树.rs
/*
 * @lc app=leetcode.cn id=617 lang=rust
 *
 * [617] 合并二叉树
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn merge_trees(root1: Option<Rc<RefCell<TreeNode>>>, root2: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        match (root1, root2) {
            (None, None) => return None,
            (some, None) | (None, some) => some,
            (Some(t1), Some(t2)) => match(t1.borrow_mut(), t2.borrow_mut()) {
                (mut t1, mut t2) => Some(Rc::new(RefCell::new(TreeNode{
                    val: t1.val + t2.val,
                    left: Self::merge_trees(t1.left.take(), t2.left.take()),
                    right: Self::merge_trees(t1.right.take(), t2.right.take()),
                })))
            } 
        }
    }
}
// @lc code=end

